### ISSUE TITLE
Adding Scope in Custom Client Does Not Work

### ISSUE THREAD

**Issue by codexboise** [alekseynemiro/nemiro.oauth.dll#5]
If you create a custom client you cannot set the **DefaultScope** variable as is done in the built in clients like the FacebookClient.cs file 

```C#
 public FacebookClient(string clientId, string clientSecret) : base
    (
      "https://www.facebook.com/dialog/oauth",
      "https://graph.facebook.com/oauth/access_token",
      clientId,
      clientSecret
    )
    {
      // scope list
      base.ScopeSeparator = ",";
      base.DefaultScope = "public_profile,email"; //,user_website,user_birthday
    }
```
This is because it is an internal property and not accessible. The other option, setting the **Scope** variable, does not work either because of the way the scope is processed in the OAuth2Client.cs file. In the code if the **DefaultScope** variable is not set then the **Scope** variable is not used.

 ```C#
        // add default scope
        string scope = this.DefaultScope;

        // add custom scope
        if (!String.IsNullOrEmpty(this.Scope))
        {
          var scopeToAdd = new List<string>();

        /**
         * This is were the problem occurs, 
         * Since we have not set the DefaultScope variable the Scope variable is not used. 
         **/
          if (!String.IsNullOrEmpty(scope))
          {
            var scope1 = scope.Split(this.ScopeSeparator.ToCharArray());
            var scope2 = this.Scope.Split(this.ScopeSeparator.ToCharArray());
            foreach (var s in scope2)
            {
              if (!scope1.Any(itm => itm.Equals(s, StringComparison.OrdinalIgnoreCase)))
              {
                scopeToAdd.Add(s);
              }
            }
            if (scopeToAdd.Count > 0) { scope += this.ScopeSeparator; }
          }
          scope += String.Join(this.ScopeSeparator, scopeToAdd.ToArray());
        }

        // add scope to url
        if (!String.IsNullOrEmpty(scope))
        {
          result += String.Format("&scope={0}", OAuthUtility.UrlEncode(scope));
        }
```

By adding an **else** to the **if** statement that is checking if the **scope** variable is empty would fix this issue.

```C#
       // add default scope
        string scope = this.DefaultScope;

        // add custom scope
        if (!String.IsNullOrEmpty(this.Scope))
        {
          var scopeToAdd = new List<string>();
          if (!String.IsNullOrEmpty(scope))
          {
            var scope1 = scope.Split(this.ScopeSeparator.ToCharArray());
            var scope2 = this.Scope.Split(this.ScopeSeparator.ToCharArray());
            foreach (var s in scope2)
            {
              if (!scope1.Any(itm => itm.Equals(s, StringComparison.OrdinalIgnoreCase)))
              {
                scopeToAdd.Add(s);
              }
            }
            if (scopeToAdd.Count > 0) { scope += this.ScopeSeparator; }
          }
          else // This is the fix
          {
            scope = Scope;
          }
          scope += String.Join(this.ScopeSeparator, scopeToAdd.ToArray());
        }

        // add scope to url
        if (!String.IsNullOrEmpty(scope))
        {
          result += String.Format("&scope={0}", OAuthUtility.UrlEncode(scope));
        }
```

**Comment by alekseynemiro** [comment]
I added the fix a problem with the **Scope**. (without release)

> If you create a custom client you cannot set the DefaultScope

It was assumed that new clients will created within the library. 
I will consider the possibility of changing this.

Thank you!

**Comment by codexboise** [comment]
How would you add a client within the library if you are using nuget to manage this package? The only way I can see being able to add it within the library is if we do not use nuget.

**Comment by alekseynemiro** [comment]
Nuget - it's just one more opportunity. 
If you make a client outside the library, it will not fully work.
Under the current architecture done differently would be difficult.

**Comment by alekseynemiro** [comment]
I made changes, with full compatibility. 
In version **1.8** will be able to create their own clients without integration with the library.

Most likely tomorrow or in Monday will be updated.

**Comment by alekseynemiro** [comment]
I released version **1.8**. Now possible to make a new clients in own program.
I added the [example](https://github.com/alekseynemiro/nemiro.oauth.dll/tree/master/src/AspNetWebFormsCustomClient) of custom clients.
